### ISSUE TITLE
Question: How do you verify the payload coming from EventsAPI?

### ISSUE THREAD

**Issue by UncleFirefox** [seratch/jslack#163]
Congratulations for the API it's very feature complete. I got one doubt on origin verification. You know you can verify requests on slack according to their practices: https://api.slack.com/docs/verifying-requests-from-slack

Maybe I just missed some part of the code but does the code for EventsAPI get the request headers and do that validation? Before calling the handlers?

Thanks!

**Comment by seratch** [comment]
The following utility is for AWS lambda but you can do similar in your applications. https://github.com/seratch/jslack/blob/1.6.0-RC2/jslack-app-backend/src/main/java/com/github/seratch/jslack/app_backend/vendor/aws/lambda/util/SlackSignatureVerifier.java

**Comment by seratch** [comment]
@UncleFirefox Do you have any further questions or comments? If not, let me close this issue within a few days.

**Comment by UncleFirefox** [comment]
How would you do it within this example?

```
public class SampleServlet extends SlackEventsApiServlet {

  @Override
  protected void setupDispatcher(EventsDispatcher dispatcher) {
    dispatcher.register(appUninstalledEventHandler);
  }
}
```

**Comment by seratch** [comment]
>It would be great if SlackEventsApiServlet passing in the proper config would do that for you before dispatching the events

Sorry, I don't understand your point here.

Generally speaking, you should verify the signature header before calling the `EventsDispatcher`.

**Comment by UncleFirefox** [comment]
Sorry for the missunderstanding:

```
protected void doPost(HttpServletRequest req, HttpServletResponse resp) throws IOException {
        String contentType = req.getHeader("Content-Type");
        if (contentType != null && contentType.toLowerCase(Locale.ENGLISH).trim().startsWith("application/json")) {
            String requestBody = req.getReader().lines().collect(Collectors.joining(System.lineSeparator()));
            JsonObject payload = GsonFactory.createSnakeCase().fromJson(requestBody, JsonElement.class).getAsJsonObject();
            String eventType = payload.get("type").getAsString();
            if (eventType != null && eventType.equals("url_verification")) {
                String challenge = payload.get("challenge").getAsString();
                // url_verification: https://api.slack.com/events/url_verification
                resp.setStatus(200);
                resp.setHeader("Content-Type", "text/plain");
                resp.getOutputStream().write(challenge.getBytes(Charset.forName("UTF-8")));
            } else {
                dispatcher.enqueue(requestBody);
                resp.setStatus(200);
            }
        } else {
            log.warn("Unexpected request detected - Content-Type: {}", req.getHeader("Content-Type"));
        }

    }
```

In the doPost method on `SlackEventsApiServlet` I think that we could verify that the token is signed correctly, potentially passing some configuration, this should be done before handling the url_verification case. Is it more clear now? How would you do it otherwise?

**Comment by seratch** [comment]
I'm not sure what you meant by the `config` here but if you mean a kind of Spring component to inject,  I don't have any plans to implement some components specific to Servlet API or Web frameworks like Spring in the short term. If you would like to have some reusable component for the purpose, you need to implement your own one.

**Comment by seratch** [comment]
If you go with Servlet, implementing a servlet filter to verify requests may be useful.

**Comment by seratch** [comment]
May I close this issue if you don't have any further comments?

**Comment by shubhvashisht** [comment]
The issue is when the slack signature is verified before calling the dopost method of the super class the body of the req object is read and is nulled out. Due to this the below line "String requestBody = req.getReader().lines().collect(Collectors.joining(System.lineSeparator()));" throws an exception.

If you could overload this method to also accept header and request body separately then it will work for the signed secret validation at the client end.

I would like to contribute for this change and you can review the pull request. Please let me know.

**Comment by seratch** [comment]
@shubhvashisht 
>The issue 

OK, you're talking about this implementation, right?
https://github.com/seratch/jslack/blob/1.7.8/jslack-app-backend/src/main/java/com/github/seratch/jslack/app_backend/events/servlet/SlackEventsApiServlet.java#L40

>when the slack signature is verified before calling the dopost method of the super class the body of the req object is read and is nulled out. Due to this the below line "String requestBody = req.getReader().lines().collect(Collectors.joining(System.lineSeparator()));" throws an exception.

I understand reading request body this way could be problematic for some cases.

>If you could overload this method to also accept header and request body separately then it will work for the signed secret validation at the client end.

I don't understand your point here yet. If you mean having a method to read and store request body values, I'm fine with it.

```java

private String cachedRequestBody;

protected String getRequestBody() {
    // read and cache the value
   return this.cachedRequestBody;
}

protected void doPost(HttpServletRequest req, HttpServletResponse resp) throws IOException {
    String contentType = req.getHeader("Content-Type");
    if (contentType != null && contentType.toLowerCase(Locale.ENGLISH).trim().startsWith("application/json")) {
        String requestBody = getRequestBody();
        // omitted
    } else {
        log.warn("Unexpected request detected - Content-Type: {}", req.getHeader("Content-Type"));
    }

}
```

**Comment by shubhvashisht** [comment]
Let me try to rephrase the issue that I am facing in the signed secret authentication for the events API.

The code you suggested above will perform the following in sequence:

1. Cache the request body 
2. Enque the request
3. Pass the control to the subscribed event handler

My concern over here is that I want to validate the incoming request even before calling this function to enque the request and handle the event. This will ensure that any malicious attack or bad request will be rejected even before it hits the doPost method for enque.

So if I retrieve the request body for signature validation then it is no longer available in the HttpServletRequest object in the dopost method and your process fails.

Below is the code sample of what I am recommending:
```Java
// Method accepting header and request body separately
// this will allow the client to do the signed secret validation before hand and not loose the jslack's capability of handling the event queue
protected void doPostWithRequestBody(Map<String, String> header, String requestBody,HttpServletResponse resp){
    String contentType = header.getHeader("Content-Type");
    if (contentType != null && contentType.toLowerCase(Locale.ENGLISH).trim().startsWith("application/json")) {
       //String requestBody = getRequestBody(); Already have the request body in the function parameter
        // omitted
    } else {
        log.warn("Unexpected request detected - Content-Type: {}", req.getHeader("Content-Type"));
    }
}

    protected void doPost(HttpServletRequest req, HttpServletResponse resp) throws IOException {
        Map<String, String> headerMap = new HashMap<String, String>();

        Enumeration headerNames = request.getHeaderNames();
        while (headerNames.hasMoreElements()) {
            String key = (String) headerNames.nextElement();
            String value = request.getHeader(key);
            headerMap.put(key, value);
        }
        String requestBody = req.getReader().lines().collect(Collectors.joining(System.lineSeparator()));
        doPostWithRequestBody(headerMap,requestBody,resp);
    }
```

Thoughts?

**Comment by seratch** [comment]
Ah you’re right. I totally thought that the servlet implementation does such validation but it doesn’t. It should be improved.

**Comment by shubhvashisht** [comment]
let me know if you would want me to work on it?

**Comment by seratch** [comment]
>let me know if you would want me to work on it?

I can work on this but if you're interested in contributing to this library, of course, I welcome your PRs!

**Comment by shubhvashisht** [comment]
submitted the PR

**Comment by seratch** [comment]
The feature requested by @shubhvashisht here has been merged in #194.  I will release a new version today. If you have further comments regarding this, feel free to reopen this issue or open a new issue for similar topics.
